src/netlink.c

```diff
@@ -0,0 +1,365 @@
+    struct nlmsghdr nlh[NLMSG_SPACE(MAX_PAYLOAD)];
+    size_t got;
+    p_timeout tm = &nl->tm;
+    int err;
+
+    memset(nlh, 0, NLMSG_SPACE(MAX_PAYLOAD));
+    nlh->nlmsg_len = NLMSG_SPACE(MAX_PAYLOAD);
+    timeout_markstart(tm);
+    err = socket_recv(&nl->fd, (char *)nlh, nlh->nlmsg_len, &got, tm);
+    if (err != IO_DONE && err != IO_CLOSED) {
+        lua_pushnil(L);
+        lua_pushstring(L, socket_strerror(err));
+        return 2;
+    }
+
+    lua_pushinteger(L, got);
+    lua_pushlstring(L, NLMSG_DATA(nlh), got);
+    return 2;
+}
+
+/*-------------------------------------------------------------------------*\
+* Receives data from a netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_receivefrom(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{unconnected}", 1);
+    struct nlmsghdr nlh[NLMSG_SPACE(MAX_PAYLOAD)];
+    struct sockaddr_nl dst;
+    size_t got;
+    p_timeout tm = &nl->tm;
+    int err;
+
+    memset(nlh, 0, NLMSG_SPACE(MAX_PAYLOAD));
+    nlh->nlmsg_len = NLMSG_SPACE(MAX_PAYLOAD);
+    socklen_t len = sizeof(dst);
+    timeout_markstart(tm);
+    err = socket_recvfrom(&nl->fd, (char *)nlh, nlh->nlmsg_len, &got, 
+            (SA *)&dst, &len, tm);
+    if (err != IO_DONE && err != IO_CLOSED) {
+        lua_pushnil(L);
+        lua_pushstring(L, socket_strerror(err));
+        return 2;
+    }
+
+    lua_pushinteger(L, got);
+    lua_pushlstring(L, NLMSG_DATA(nlh), got);
+    lua_pushinteger(L, nlh->nlmsg_pid);
+    return 3;
+}
+
+/*-------------------------------------------------------------------------*\
+* Select support methods
+\*-------------------------------------------------------------------------*/
+static int meth_getfd(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkgroup(L, "netlink{any}", 1);
+    lua_pushnumber(L, (int)nl->fd);
+    return 1;
+}
+
+/* this is very dangerous, but can be handy for those that are brave enough */
+static int meth_setfd(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkgroup(L, "netlink{any}", 1);
+    nl->fd = (t_socket)luaL_checknumber(L, 2);
+    return 0;
+}
+
+static int meth_getsockpid(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkgroup(L, "netlink{any}", 1);
+    lua_pushinteger(L, nl->srcpid);
+    return 1;
+}
+
+/*-------------------------------------------------------------------------*\
+* Binds an object to an address
+\*-------------------------------------------------------------------------*/
+static const char *netlink_trybind(p_netlink nl, int grp) {
+    struct sockaddr_nl addr;
+    int err;
+
+    memset(&addr, 0, sizeof(addr));
+    addr.nl_family = AF_NETLINK;
+    addr.nl_groups = grp;	
+    addr.nl_pid = nl->srcpid;
+    getpeername(nl->fd, (SA *)&addr, (socklen_t *)sizeof(addr));
+    err = socket_bind(&nl->fd, (SA *)&addr, sizeof(addr));
+    return socket_strerror(err);
+}
```

Comment:
Why did you separate this code in another function? It is only used in one place and there is no more complex than the code in most other functions.

Reply:
I followed the style used by Diego. He uses a try_ function for both bind and connect, so I did the same, except there's no connect here.

Reply:
So, if Diego's code jumps off a bridge... ah, never mind.